File: benchmarks/tasks/stack_calc/stack_calc.hs
-- SPDX-License-Identifier: MIT-0
module Main where

data Op = Push Int | Add | Mul

eval :: [Op] -> [Int] -> Int
eval [] (x:_)             = x
eval (Push n : ops) stack = eval ops (n : stack)
eval (Add : ops) (a:b:s)  = eval ops ((b + a) : s)
eval (Mul : ops) (a:b:s)  = eval ops ((b * a) : s)
eval _ _                  = error "invalid"

main :: IO ()
main = print (eval [Push 3, Push 4, Add, Push 5, Mul] [])

File: benchmarks/tasks/stack_calc/stack_calc.py
# SPDX-License-Identifier: MIT-0

def eval_rpn(ops):
    stack = []
    for op in ops:
        if isinstance(op, tuple) and op[0] == "push":
            stack.append(op[1])
        elif op == "add":
            b, a = stack.pop(), stack.pop()
            stack.append(a + b)
        elif op == "mul":
            b, a = stack.pop(), stack.pop()
            stack.append(a * b)
    return stack[0]

result = eval_rpn([("push", 3), ("push", 4), "add", ("push", 5), "mul"])
print(result)
